<a href="https://colab.research.google.com/github/falyseck/Multimodel_Data_Prepocessing/blob/main/model_training.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Multimodel Logic
This project uses a multimodal approach, combining three different machine learning models:

- A facial recognition model for identifying

the user

- A voice verification model for confirming identity for transaction

- A product recommendation model for predicting user preferences

The system follows a sequential decision process:

1. The facial recognition model predicts the user’s identity

2. The voice model independently verifies the user

If both predictions match, the user is authenticated

3. Once authenticated, the product recommendation model is used to suggest a product

This approach improves system security by requiring two independent biometric verifications before allowing a transaction.

# Face recognition

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score
import joblib

# Load image features
image_df = pd.read_csv('image_features.csv')
X = image_df.drop(columns= ['person','image'], axis=1)
y = image_df['person']

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model
face_model = RandomForestClassifier(n_estimators=100, random_state=42)
face_model.fit(X_train, y_train)

# Evaluate
y_pred = face_model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("F1-Score:", f1_score(y_test, y_pred, average='weighted'))

# Save model
joblib.dump(face_model, 'face_model.pkl')

Accuracy: 1.0
F1-Score: 1.0


['face_model.pkl']

## Model Evaluation Summary

### Face Recognition Model
The `face_model` (RandomForestClassifier) was trained and evaluated on image features. The evaluation metrics are:
- **Accuracy:** 1.0
- **F1-Score (weighted):** 1.0

*(Note: For RandomForestClassifiers, a 'loss' metric is not typically reported as it is for deep learning models; accuracy and F1-score are standard evaluation metrics for classification tasks.)*

# Voiceprint verification

In [ ]:
# Load audio features
audio_df = pd.read_csv('audio_features.csv')
X = audio_df.drop('person', axis=1)
y = audio_df['person']

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model
voice_model = RandomForestClassifier(n_estimators=100, random_state=42)
voice_model.fit(X_train, y_train)

# Evaluate
y_pred = voice_model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("F1-Score:", f1_score(y_test, y_pred, average='weighted'))

# Save model
joblib.dump(voice_model, 'voice_model.pkl')

Accuracy: 1.0
F1-Score: 1.0


['voice_model.pkl']

### Voiceprint Verification Model
The `voice_model` (RandomForestClassifier) was trained and evaluated on audio features. The evaluation metrics are:
- **Accuracy:** 1.0
- **F1-Score (weighted):** 1.0


# Product recommendation

In [ ]:
# Load merged dataset
merged_df = pd.read_csv('merged_customer_data.csv')

# Define features and target
X = merged_df.drop(['product_category', 'customer_id'], axis=1)
y = merged_df['product_category']

# Drop non-numeric columns from X
non_numeric_cols = X.select_dtypes(include=['object']).columns
X = X.drop(columns=non_numeric_cols)

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model
product_model = RandomForestClassifier(n_estimators=200, random_state=42)
product_model.fit(X_train, y_train)

# Evaluate
y_pred = product_model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("F1-Score:", f1_score(y_test, y_pred, average='weighted'))

# Save model
joblib.dump(product_model, 'product_model.pkl')

Accuracy: 0.26666666666666666
F1-Score: 0.2500552079499448


['product_model.pkl']

### Product Recommendation Model
The `product_model` (RandomForestClassifier) was trained and evaluated on merged customer data. The evaluation metrics are:
- **Accuracy:** 0.26666666666666666
- **F1-Score (weighted):** 0.2500552079499448

**The relatively low accuracy and F1-score for this model are likely due to the small size of the dataset used for training and evaluation, which might not be sufficient for the model to learn complex patterns effectively.**